### Data Warehouse and data mining systems #2

In [1]:
import pandas as pd
import numpy as np 
import ast

DATA_DIR="datasets\\"

In [2]:
data1 = pd.read_csv("".join([DATA_DIR,"SpotifyTop200songs2017-2021\\data.csv"]), sep="#")
data1 = data1[data1['Track Name'].notna()] # deleting songs without names
data2 = pd.read_excel("".join([DATA_DIR,"SpotifyTop200songs2017-2021\Spotify 2010 - 2019 Top 100 Songs.xlsx"]))
data2["rank"] = np.tile(np.arange(1, 101), 10)
[i for i in range(1, 101)]*10
data2 = data2[data2['title'].notna()] # deleting songs without names
# print(data1.head(5))
# print('-'*120)
# print(data2.head(5))
print('-'*120)
data_merged = data1.merge(data2, left_on=['Track Name', 'Artist'], right_on=['title', 'artist'], how='outer')
print(data_merged.shape)
print(data_merged.columns)
print(data_merged.head(10))
print(data_merged.isnull().values.any())

------------------------------------------------------------------------------------------------------------------------
(325532, 24)
Index(['Position', 'Track Name', 'Artist', 'Streams', 'Date', 'Genre', 'title',
       'artist', 'top genre', 'year released', 'added', 'bpm', 'nrgy', 'dnce',
       'dB', 'live', 'val', 'dur', 'acous', 'spch', 'pop', 'top year',
       'artist type', 'rank'],
      dtype='object')
   Position Track Name      Artist    Streams        Date  \
0       1.0    Starboy  The Weeknd  3135625.0  2017-01-01   
1       1.0    Starboy  The Weeknd  3342769.0  2017-01-02   
2       1.0    Starboy  The Weeknd  3563076.0  2017-01-03   
3       1.0    Starboy  The Weeknd  3619247.0  2017-01-04   
4       1.0    Starboy  The Weeknd  3579206.0  2017-01-05   
5       3.0    Starboy  The Weeknd  3590301.0  2017-01-06   
6       3.0    Starboy  The Weeknd  3460752.0  2017-01-07   
7       3.0    Starboy  The Weeknd  3127938.0  2017-01-08   
8       3.0    Starboy  The Weeknd


#









<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Position</th>
      <th>Track Name</th>
      <th>Artist</th>
      <th>Streams</th>
      <th>Date</th>
      <th>Genre</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>Starboy</td>
      <td>The Weeknd</td>
      <td>3135625</td>
      <td>2017-01-01</td>
      <td>['canadian pop', 'canadian contemporary r&amp;b', 'pop']</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>Closer</td>
      <td>The Chainsmokers</td>
      <td>3015525</td>
      <td>2017-01-01</td>
      <td>['pop', 'pop dance', 'tropical house', 'edm', 'electropop', 'dance pop']</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3</td>
      <td>Let Me Love You</td>
      <td>DJ Snake</td>
      <td>2545384</td>
      <td>2017-01-01</td>
      <td>['pop', 'electronic trap', 'dance pop', 'edm', 'pop dance', 'pop rap']</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>Rockabye (feat. Sean Paul &amp; Anne-Marie)</td>
      <td>Clean Bandit</td>
      <td>2356604</td>
      <td>2017-01-01</td>
      <td>['pop', 'uk dance', 'dance pop', 'uk funky', 'tropical house', 'pop dance', 'post-teen pop', 'edm']</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5</td>
      <td>One Dance</td>
      <td>Drake</td>
      <td>2259887</td>
      <td>2017-01-01</td>
      <td>['toronto rap', 'canadian pop', 'canadian hip hop', 'rap', 'pop rap', 'hip hop']</td>
    </tr>
  </tbody>
</table>

In [3]:
merged = data_merged.loc[:,["title", "Track Name"]].isnull().values.all(axis=1)
null_indexes = [i for i in range(len(merged)) if merged[i]]
data_merged.loc[null_indexes,:].head(5) # our data is properly merged

Empty DataFrame
Columns: [Position, Track Name, Artist, Streams, Date, Genre, title, artist, top genre, year released, added, bpm, nrgy, dnce, dB, live, val, dur, acous, spch, pop, top year, artist type, rank]
Index: []

[0 rows x 24 columns]

In [4]:
merged = data_merged.loc[:,["Track Name"]].isnull().values.all(axis=1)
null_indexes = [i for i in range(len(merged)) if merged[i]]
not_null_indexes = [i for i in range(len(merged)) if not merged[i]]
data_merged.loc[null_indexes,:].head(5)

Position Track Name Artist  Streams Date Genre  \
324917       NaN        NaN    NaN      NaN  NaN   NaN   
324918       NaN        NaN    NaN      NaN  NaN   NaN   
324919       NaN        NaN    NaN      NaN  NaN   NaN   
324920       NaN        NaN    NaN      NaN  NaN   NaN   
324921       NaN        NaN    NaN      NaN  NaN   NaN   

                                                title      artist  \
324917                  STARSTRUKK (feat. Katy Perry)       3OH!3   
324918                    My First Kiss (feat. Ke$ha)       3OH!3   
324919                                I Need A Dollar  Aloe Blacc   
324920  Airplanes (feat. Hayley Williams of Paramore)       B.o.B   
324921              Nothin' on You (feat. Bruno Mars)       B.o.B   

          top genre  year released  ...   dB  live   val    dur  acous  spch  \
324917    dance pop         2009.0  ... -6.0  23.0  23.0  203.0    0.0   6.0   
324918    dance pop         2010.0  ... -4.0  36.0  83.0  192.0    1.0   8.0   
324919     pop soul         2010.0  ... -7.0   9.0  96.0  243.0   20.0   3.0   
324920  atl hip hop         2010.0  ... -4.0   4.0  38.0  180.0   11.0  12.0   
324921  atl hip hop         2010.0  ... -6.0   9.0  74.0  268.0   39.0   5.0   

         pop  top year  artist type  rank  
324917  70.0    2010.0          Duo   1.0  
324918  68.0    2010.0          Duo   2.0  
324919  72.0    2010.0         Solo   3.0  
324920  80.0    2010.0         Solo   4.0  
324921  79.0    2010.0         Solo   5.0  

[5 rows x 24 columns]

In [5]:
data_merged["Track Name"].fillna(data_merged.title, inplace=True)  
data_merged["Artist"].fillna(data_merged.artist, inplace=True)
data_merged["Date"].fillna(data_merged.added, inplace=True)
data_merged["Position"].fillna(data_merged["rank"], inplace=True)
data_merged.drop(columns=['title', 'artist', 'added', "rank"], inplace=True)
data_merged.loc[null_indexes,:].head(5)

Position                                     Track Name      Artist  \
324917       1.0                  STARSTRUKK (feat. Katy Perry)       3OH!3   
324918       2.0                    My First Kiss (feat. Ke$ha)       3OH!3   
324919       3.0                                I Need A Dollar  Aloe Blacc   
324920       4.0  Airplanes (feat. Hayley Williams of Paramore)       B.o.B   
324921       5.0              Nothin' on You (feat. Bruno Mars)       B.o.B   

        Streams        Date Genre    top genre  year released    bpm  nrgy  \
324917      NaN  2022‑02‑17   NaN    dance pop         2009.0  140.0  81.0   
324918      NaN  2022‑02‑17   NaN    dance pop         2010.0  138.0  89.0   
324919      NaN  2022‑02‑17   NaN     pop soul         2010.0   95.0  48.0   
324920      NaN  2022‑02‑17   NaN  atl hip hop         2010.0   93.0  87.0   
324921      NaN  2022‑02‑17   NaN  atl hip hop         2010.0  104.0  85.0   

        dnce   dB  live   val    dur  acous  spch   pop  top year artist type  
324917  61.0 -6.0  23.0  23.0  203.0    0.0   6.0  70.0    2010.0         Duo  
324918  68.0 -4.0  36.0  83.0  192.0    1.0   8.0  68.0    2010.0         Duo  
324919  84.0 -7.0   9.0  96.0  243.0   20.0   3.0  72.0    2010.0        Solo  
324920  66.0 -4.0   4.0  38.0  180.0   11.0  12.0  80.0    2010.0        Solo  
324921  69.0 -6.0   9.0  74.0  268.0   39.0   5.0  79.0    2010.0        Solo

In [6]:
data_merged.loc[null_indexes,:].shape

(615, 20)

#### Setting genres

In [7]:
genres = [[], [], []]

for genre_list, genre in zip(data_merged["Genre"], data_merged["top genre"]):
    if pd.isna(genre_list):
        genres[0].append(genre)
        genres[1].append(np.nan)
        genres[2].append(np.nan)
    else:
        genre_list = ast.literal_eval(genre_list)
        for i in range(min(len(genre_list), 3)):
            genres[i].append(genre_list[i])
        for i in range(len(genre_list), 3):
            genres[i].append(np.nan)

In [8]:
data_merged["genre_1"] = genres[0]
data_merged["genre_2"] = genres[1]
data_merged["genre_3"] = genres[2]

In [9]:
data_merged.drop(columns=["Genre", "top genre"], inplace=True)
data_merged.loc[null_indexes,:].head(5)

Position                                     Track Name      Artist  \
324917       1.0                  STARSTRUKK (feat. Katy Perry)       3OH!3   
324918       2.0                    My First Kiss (feat. Ke$ha)       3OH!3   
324919       3.0                                I Need A Dollar  Aloe Blacc   
324920       4.0  Airplanes (feat. Hayley Williams of Paramore)       B.o.B   
324921       5.0              Nothin' on You (feat. Bruno Mars)       B.o.B   

        Streams        Date  year released    bpm  nrgy  dnce   dB  ...   val  \
324917      NaN  2022‑02‑17         2009.0  140.0  81.0  61.0 -6.0  ...  23.0   
324918      NaN  2022‑02‑17         2010.0  138.0  89.0  68.0 -4.0  ...  83.0   
324919      NaN  2022‑02‑17         2010.0   95.0  48.0  84.0 -7.0  ...  96.0   
324920      NaN  2022‑02‑17         2010.0   93.0  87.0  66.0 -4.0  ...  38.0   
324921      NaN  2022‑02‑17         2010.0  104.0  85.0  69.0 -6.0  ...  74.0   

          dur  acous  spch   pop  top year  artist type      genre_1 genre_2  \
324917  203.0    0.0   6.0  70.0    2010.0          Duo    dance pop     NaN   
324918  192.0    1.0   8.0  68.0    2010.0          Duo    dance pop     NaN   
324919  243.0   20.0   3.0  72.0    2010.0         Solo     pop soul     NaN   
324920  180.0   11.0  12.0  80.0    2010.0         Solo  atl hip hop     NaN   
324921  268.0   39.0   5.0  79.0    2010.0         Solo  atl hip hop     NaN   

       genre_3  
324917     NaN  
324918     NaN  
324919     NaN  
324920     NaN  
324921     NaN  

[5 rows x 21 columns]

In [10]:
artist_genre = pd.concat([
    data_merged.loc[:, ['Artist','genre_1']], 
    data_merged.loc[:, ['Artist','genre_2']].rename(columns={'genre_2':'genre_1'}),
    data_merged.loc[:, ['Artist','genre_3']].rename(columns={'genre_3':'genre_1'})])
artist_genre.dropna(inplace=True)
artist_genre = artist_genre.groupby(['Artist', 'genre_1']).size().reset_index(name="occurences")
artist_genre.head(6)

Artist         genre_1  occurences
0       *NSYNC        boy band          18
1       *NSYNC       dance pop          18
2       *NSYNC   post-teen pop          18
3  13 Organisé  french hip hop          17
4  13 Organisé     pop urbaine          17
5  13 Organisé    rap francais          17

In [11]:
best_genre = artist_genre.loc[artist_genre.groupby('Artist')["occurences"].idxmax(), ["Artist", "genre_1"]]
best_genre.rename(columns={'genre_1':'main_genre'}, inplace=True)
best_genre.head(5)

Artist      main_genre
0              *NSYNC        boy band
3         13 Organisé  french hip hop
6   187 Strassenbande  german hip hop
8            2 Chainz     atl hip hop
11          21 Savage     atl hip hop

#### SongProps pd df creation

In [12]:
SongProps = data_merged.loc[:, ['Track Name', 'Artist', 'nrgy', 'dnce', 'dB', 'live', 'bpm', 'val', 'dur', 'acous', 'spch', 'pop', 'genre_1', 'genre_2', 'genre_3', 'year released', 'artist type']]
SongProps.drop_duplicates(ignore_index=True, inplace=True)
SongProps = SongProps.reset_index(level=0)
SongProps.rename(columns = {"index":"song_prop_id", "artist type": "ensemble"}, inplace = True)
SongProps.head(5)

song_prop_id                               Track Name            Artist  \
0             0                                  Starboy        The Weeknd   
1             1                                   Closer  The Chainsmokers   
2             2                          Let Me Love You          DJ Snake   
3             3  Rockabye (feat. Sean Paul & Anne-Marie)      Clean Bandit   
4             4                                One Dance             Drake   

   nrgy  dnce   dB  live    bpm   val    dur  acous  spch   pop       genre_1  \
0  59.0  68.0 -7.0  14.0  186.0  49.0  230.0   14.0  28.0  88.0  canadian pop   
1  52.0  75.0 -6.0  11.0   95.0  66.0  245.0   41.0   3.0  88.0           pop   
2  72.0  65.0 -5.0  14.0  100.0  16.0  206.0    9.0   3.0  86.0           pop   
3  76.0  72.0 -4.0  18.0  102.0  74.0  251.0   41.0   5.0  80.0           pop   
4   NaN   NaN  NaN   NaN    NaN   NaN    NaN    NaN   NaN   NaN   toronto rap   

                     genre_2           genre_3  year released    ensemble  
0  canadian contemporary r&b               pop         2016.0        Solo  
1                  pop dance    tropical house         2016.0         Duo  
2            electronic trap         dance pop         2016.0        Solo  
3                   uk dance         dance pop         2016.0  Band/Group  
4               canadian pop  canadian hip hop            NaN         NaN

In [13]:
SongPropsIdx = SongProps.loc[:, ['song_prop_id', 'Track Name', 'Artist']]
SongProps    = SongProps.loc[:, ['song_prop_id', 'nrgy', 'dnce', 'dB', 'live', 'bpm', 'val', 'dur', 'acous', 'spch', 'pop', 'genre_1', 'genre_2', 'genre_3', 'year released', 'ensemble']]
SongProps.head(5)

song_prop_id  nrgy  dnce   dB  live    bpm   val    dur  acous  spch   pop  \
0             0  59.0  68.0 -7.0  14.0  186.0  49.0  230.0   14.0  28.0  88.0   
1             1  52.0  75.0 -6.0  11.0   95.0  66.0  245.0   41.0   3.0  88.0   
2             2  72.0  65.0 -5.0  14.0  100.0  16.0  206.0    9.0   3.0  86.0   
3             3  76.0  72.0 -4.0  18.0  102.0  74.0  251.0   41.0   5.0  80.0   
4             4   NaN   NaN  NaN   NaN    NaN   NaN    NaN    NaN   NaN   NaN   

        genre_1                    genre_2           genre_3  year released  \
0  canadian pop  canadian contemporary r&b               pop         2016.0   
1           pop                  pop dance    tropical house         2016.0   
2           pop            electronic trap         dance pop         2016.0   
3           pop                   uk dance         dance pop         2016.0   
4   toronto rap               canadian pop  canadian hip hop            NaN   

     ensemble  
0        Solo  
1         Duo  
2        Solo  
3  Band/Group  
4         NaN

In [14]:
data_merged = data_merged.merge(SongPropsIdx, how='left', left_on=['Track Name', 'Artist'], right_on=['Track Name', 'Artist'])
data_merged.drop(columns= ['nrgy', 'dnce', 'dB', 'live', 'bpm', 'genre_1', 'genre_2', 'genre_3', 'year released', 'val', 'dur', 'acous', 'spch', 'pop', 'artist type'], inplace=True)
data_merged.head(5)

Position Track Name      Artist    Streams        Date  top year  \
0       1.0    Starboy  The Weeknd  3135625.0  2017-01-01    2016.0   
1       1.0    Starboy  The Weeknd  3342769.0  2017-01-02    2016.0   
2       1.0    Starboy  The Weeknd  3563076.0  2017-01-03    2016.0   
3       1.0    Starboy  The Weeknd  3619247.0  2017-01-04    2016.0   
4       1.0    Starboy  The Weeknd  3579206.0  2017-01-05    2016.0   

   song_prop_id  
0             0  
1             0  
2             0  
3             0  
4             0

#### Artist pd df creation

In [15]:
data_merged = data_merged.merge(best_genre, how='left')
Artists = data_merged.loc[:, ["Artist", "main_genre"]].drop_duplicates(ignore_index=True)
Artists = Artists.reset_index(level=0)
Artists.rename(columns = {'Artist':'artist_name', "index":"artist_id"}, inplace = True)
Artists

artist_id       artist_name                 main_genre
0             0        The Weeknd  canadian contemporary r&b
1             1  The Chainsmokers                        pop
2             2          DJ Snake                  dance pop
3             3      Clean Bandit                  dance pop
4             4             Drake           canadian hip hop
...         ...               ...                        ...
1283       1283               M.O                  afroswing
1284       1284                MK                  dance pop
1285       1285         AJ Tracey                      grime
1286       1286               NSG             afro dancehall
1287       1287             Wiley                      grime

[1288 rows x 3 columns]

In [16]:
data_merged = data_merged.merge(Artists.loc[:, ['artist_name', 'artist_id']], how='left', left_on='Artist', right_on='artist_name')
data_merged.drop(columns= ['artist_name', 'Artist', 'main_genre', 'top year'], inplace=True)
data_merged.head(5) # potential revenue db?
# what about the rest of properties here?

Position Track Name    Streams        Date  song_prop_id  artist_id
0       1.0    Starboy  3135625.0  2017-01-01             0          0
1       1.0    Starboy  3342769.0  2017-01-02             0          0
2       1.0    Starboy  3563076.0  2017-01-03             0          0
3       1.0    Starboy  3619247.0  2017-01-04             0          0
4       1.0    Starboy  3579206.0  2017-01-05             0          0

In [17]:
Artists.shape

(1288, 3)